src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -167,44 +166,29 @@ public IImmutableSet<ProjectId> GetProjectsThatDirectlyDependOnThisProject(Proje
                 throw new ArgumentNullException(nameof(projectId));
             }
 
-            if (_lazyReverseReferencesMap == null)
-            {
-                using (_dataLock.DisposableWait())
-                {
-                    return this.GetProjectsThatDirectlyDependOnThisProject_NoLock(projectId);
-                }
-            }
-            else
-            {
-                // okay, because its only ever going to be computed and assigned once
-                return this.GetProjectsThatDirectlyDependOnThisProject_NoLock(projectId);
-            }
+            return this.GetProjectsThatDirectlyDependOnThisProject_NoLock(projectId);
         }
 
         private ImmutableHashSet<ProjectId> GetProjectsThatDirectlyDependOnThisProject_NoLock(ProjectId projectId)
         {
-            if (_lazyReverseReferencesMap == null)
-            {
-                _lazyReverseReferencesMap = this.ComputeReverseReferencesMap();
-                ValidateReverseReferences(_projectIds, _referencesMap, _lazyReverseReferencesMap);
-            }
+            return ImmutableInterlocked.GetOrAdd(
+                ref _lazyReverseReferencesMap,
+                projectId,
+                static (projectId, referencesMap) => ComputeReverseReferences(referencesMap, projectId),
+                _referencesMap);
 
-            return _lazyReverseReferencesMap.GetValueOrDefault(projectId, ImmutableHashSet<ProjectId>.Empty);
-        }
+            static ImmutableHashSet<ProjectId> ComputeReverseReferences(ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> referencesMap, ProjectId requestedProjectId)
+            {
+                using var _ = PooledHashSet<ProjectId>.GetInstance(out var hashSet);
 
-        private ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeReverseReferencesMap()
-        {
-            var reverseReferencesMap = new Dictionary<ProjectId, HashSet<ProjectId>>();
+                foreach (var (projectId, references) in referencesMap)
+                {
+                    if (references.Contains(requestedProjectId))
+                        hashSet.Add(projectId);
+                }
 
-            foreach (var (projectId, references) in _referencesMap)
-            {
-                foreach (var referencedId in references)
-                    reverseReferencesMap.MultiAdd(referencedId, projectId);
+                return hashSet.Count > 0 ? hashSet.ToImmutableHashSet() : ImmutableHashSet<ProjectId>.Empty;
```

Comment:
is this not just how ToImmutableHashSet would already work?

Reply:
Not from what I can see when inspecting the code.

Reply:
Well that's a tragedy :-)

Reply:
Agreed